core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointStatus.scala

```diff
@@ -0,0 +1,310 @@
+  }
+
+  object Deviation {
+    final private[this] case class DeviationImpl(override val mapping: List[CodePoint])
+        extends Deviation
+
+    /**
+     * Status for a deviation code point which is ignored, e.g. not dropped, in the UTS-46
+     * mapping step.
+     */
+    val ignored: Deviation = DeviationImpl(Nil)
+
+    /**
+     * Create a [[Deviation]] status given a `NonEmptyList` of [[CodePoint]]
+     *
+     * @note
+     *   If you wish to create a status for a deviation code point which is not mapped, e.g.
+     *   dropped/ignored, please use [[#ignored]].
+     */
+    def of(mapping: NonEmptyList[CodePoint]): Deviation =
+      DeviationImpl(mapping.toList)
+
+    /**
+     * Create a [[Deviation]] status given a [[CodePoint]].
+     */
+    def one(mapping: CodePoint): Deviation =
+      of(NonEmptyList.one(mapping))
+  }
+
+  /**
+   * Status for a disallowed code point. Disallowed code points will result in an error when
+   * they occur in the input to the UTS-46 mapping step.
+   */
+  sealed abstract class Disallowed extends CodePointStatus {
+    final override def toString: String = "Disallowed"
+  }
+
+  object Disallowed {
+    private[this] case object DisallowedImpl extends Disallowed
+
+    val instance: Disallowed = DisallowedImpl
+  }
+
+  /**
+   * Status for a code point that is disallowed if `useStdASCIIRules` is enabled (recommend by
+   * UTS-46), but is valid if `useStdASCIIRules` is disabled.
+   */
+  sealed abstract class Disallowed_STD3_Valid extends CodePointStatus {
+    final override def toString: String = "Disallowed_STD3_Valid"
+  }
+
+  object Disallowed_STD3_Valid {
+    private[this] case object Disallowed_STD3_ValidImpl extends Disallowed_STD3_Valid
+
+    val instance: Disallowed_STD3_Valid = Disallowed_STD3_ValidImpl
+  }
+
+  /**
+   * Status for a code point that is disallowed if `useStdASCIIRules` is enabled (recommend by
+   * UTS-46), but is mapped to one or more code points if `useStdASCIIRules` is disabled.
+   */
+  sealed abstract class Disallowed_STD3_Mapped extends CodePointStatus {
+    def mapping: NonEmptyList[CodePoint]
+
+    final override def toString: String = s"Disallowed_STD3_Mapped(mapping = ${mapping})"
+  }
+
+  object Disallowed_STD3_Mapped {
+    final private[this] case class Disallowed_STD3_MappedImpl(
+        override val mapping: NonEmptyList[CodePoint])
+        extends Disallowed_STD3_Mapped
+
+    def of(mapping: NonEmptyList[CodePoint]): Disallowed_STD3_Mapped =
+      Disallowed_STD3_MappedImpl(mapping)
+
+    def one(mapping: CodePoint): Disallowed_STD3_Mapped =
+      of(NonEmptyList.one(mapping))
+  }
+
+  /**
+   * Special ADT constructor for binary compatibility.
+   *
+   * This type is present to allow greater latitude in case a binary breaking change is required
+   * in this ADT. It is not expected that it will ever be used.
+   */
+  sealed abstract class Unknown extends CodePointStatus
```

Comment:
I understand the goal here about leaving the door open but I am not sure if this is the right way to achieve that. In the unlikely event we have to add another member to the ADT would it have to extend `Unknown`?

I think we either want to unseal `CodePointStatus` or make `Unknown` private. Either way users should be unable to prove that they've covered all the cases—which is what we want.

But honestly it's hard to reason about this. Would be interesting to add a compile-only test outside of the idna4s package that demonstrates that users are forced to consider a `case _ =>` for any total pattern match on `CodePointStatus`.

Reply:
So, after reviewing UTS-46 some more, I think we can just remove `Unknown`. They don't explicitly call out that they will never add new status types, but they do say that they will only ever move `disallowed` code points into the other status types.

I think that implies pretty strongly we don't need to worry about them adding new status types. If we do end up in that position, then we can create a new ADT and new functions, which would be very annoying, but given I'm pretty sure that won't happen, I'm okay with that risk.

Thoughts?

Reply:
👍 to removing.